cilium/CFP-34577-per-service-lb-algorithm-selection.md

```diff
@@ -0,0 +1,144 @@
+# CFP-34577: Per-service Load Balancing Algorithm Selection
+
+**SIG: SIG-DATAPATH**  
+**Begin Design Discussion: 2024-08-28**  
+**Cilium Release:** 1.17  
+**Authors:** Aleksander Mistewicz \<amistewicz@google.com\>
+
+## Summary
+
+Add annotation `service.cilium.io/lb-algorithm` support to Cilium which will allow a user to select the load balancing algorithm on a per-service basis.
+
+## Motivation
+
+Before this CFP, a user needs to make a choice of load-balancing algorithm to  
+use in their cluster. There is no one-size-fits-all choice.
+
+`cilium-agent` can be configured by setting `bpf-lb-algorithm` to use one of the following load balancing algorithms:
+
+1. Maglev:  
+   1. compute [a hash](https://github.com/cilium/cilium/blob/v1.16.0-rc.2/bpf/lib/hash.h#L9-L17) based on Source Address, Source Port, Destination Port and Protocol  
+   2. select a `backend_id` from a list of repeated `backend_id`s with deterministic sequence and distribution by weight  
+2. Random:  
+   1. select a `backend_id` from a list of repeated `lb_{4,6}_service`s  
+      where each `backend_id` occurs once
+
+| Algorithm | Static backend selection | Memory use (per service[^1]) | Memory use (total[^2]) |
+| :---- | :---- | :---- | :---- |
+| Maglev | \+ (as long as backends list is the same) | [65.5 KB \= 4B x 16381](https://github.com/cilium/cilium/blob/v1.16.0-rc.2/bpf/lib/lb.h#L168) (constant) | 655 MB (10 000 services) |
+| Random | \- | 120 KB \= [12 B](https://github.com/cilium/cilium/blob/v1.16.0-rc.2/bpf/lib/common.h#L1053-L1067) x backends | 3 MB (250 000 endpoints) |
+
+Memory footprint is smaller for `maglev` when there are between `5461` and `16381` endpoints in a service. There were two attempts to improve to improve `maglev`’s memory use in [cilium\#14396](https://github.com/cilium/cilium/issues/14396) and [cilium\#15019](https://github.com/cilium/cilium/pull/15019).
+
+Similar requests that failed to gain traction:
+
+1. [cilium\#8723](https://github.com/cilium/cilium/issues/8723) \- algorithm selection in service annotation  
+2. [cilium\#26264](https://github.com/cilium/cilium/issues/26264) \- more load-balancing algorithms
+
+`kube-router` lets users [select a load-balancing algorithm based on service annotation](https://github.com/cloudnativelabs/kube-router/blob/v2.1.3/docs/user-guide.md#load-balancing-scheduling-algorithms).  
+At the time of this writing, there are [14 algorithms](https://keepalived-pqa.readthedocs.io/en/latest/scheduling_algorithms.html) to choose from.
+
+## Goals
+
+1. Support multiple load-balancing algorithms on a single node
+
+## Non-Goals
+
+1. Add support for new (other than `maglev`/`random`) load-balancing algorithms
+
+## Proposal
+
+Add a new value for `bpf-lb-algorithm` and a new `service.cilium.io/scheduler` annotation which will initially accept `maglev` or `random` values.
+
+### Overview
+
```

Comment:
This text doesn't match the latest of the Google Doc. Is it up to date? Reviewers have approved the google doc but not this text. Maybe the text here just needs updating.